database/intializeDB.go

```diff
@@ -3,49 +3,146 @@ package database
-	if connectionString == "" {
-		return nil, errors.Errors{StatusCode: 500, Code: "Invalid Dialect", Reason: "invalid dialect given"}
+	maxIdleConnections, err = strconv.Atoi(golf.Config.Get(prefix + "DB_MAX_IDLE_CONNECTIONS"))
+	if err != nil {
+		maxIdleConnections = 5
 	}
 
-	db, err := sql.Open(c.dialect, connectionString)
+	maxConnections, err = strconv.Atoi(golf.Config.Get(prefix + "DB_MAX_CONNECTIONS"))
 	if err != nil {
-		configs.Log.Errorf("Failed to initialize the DB, Error:%v", err)
-		return nil, err
+		maxConnections = 20
 	}
 
-	err = db.Ping()
+	connectionMaxLifeTime, err = strconv.Atoi(golf.Config.Get(prefix + "DB_CONNECTIONS_MAX_LIFETIME"))
 	if err != nil {
-		configs.Log.Errorf("Failed to initialize the DB, Error:%v", err)
-		return nil, err
+		connectionMaxLifeTime = 15
 	}
 
-	configs.Log.Info("database is connected successfully")
-	return db, nil
+	idleConnectionTimeout, err = strconv.Atoi(golf.Config.Get(prefix + "DB_IDLE_CONNECTION_TIMEOUT"))
+	if err != nil {
+		idleConnectionTimeout = 10
+	}
+
+	monitoring, err = strconv.ParseBool(golf.Config.Get(prefix + "DB_MONITORING"))
+	if err != nil {
+		monitoring = false
+	}
+
+	retry, err = strconv.Atoi(golf.Config.Get(prefix + "DB_RETRY_COUNT"))
+	if err != nil {
+		retry = 5
+	}
+
+	retryTime, err = strconv.Atoi(golf.Config.Get(prefix + "DB_RETRY_DURATION"))
+	if err != nil {
+		retryTime = 5
+	}
+
+	c := dbConfig{
+		host:                  golf.Config.Get(prefix + "DB_HOST"),
+		password:              golf.Config.Get(prefix + "DB_PASSWORD"),
+		user:                  golf.Config.Get(prefix + "DB_USER"),
+		port:                  golf.Config.Get(prefix + "DB_PORT"),
+		dialect:               golf.Config.Get(prefix + "DB_DIALECT"),
+		dbName:                golf.Config.Get(prefix + "DB_NAME"),
+		maxOpenConns:          maxConnections,
+		maxIdleConns:          maxIdleConnections,
+		connMaxLifeTime:       connectionMaxLifeTime,
+		idleConnectionTimeout: idleConnectionTimeout,
+		monitoringEnable:      monitoring,
+	}
+
+	if c.host != "" && c.port != "" && c.user != "" && c.password != "" && c.dialect != "" {
+		db, err := establishDBConnection(golf.Logger, c)
+		if err == nil {
+			db.SetMaxOpenConns(c.maxOpenConns)
+			db.SetMaxIdleConns(c.maxIdleConns)
+			db.SetConnMaxLifetime(time.Minute * time.Duration(c.maxOpenConns))
+			db.SetConnMaxIdleTime(time.Minute * time.Duration(c.idleConnectionTimeout))
+
+			golf.Conn = db
+
+			go monitoringDB(golf, c, retry, retryTime)
+		}
+	}
 }
 
-func generateConnectionString(c config) string {
+func monitoringDB(golf model.GoLF, c dbConfig, retry, retryTime int) {
+	ticker := time.NewTicker(time.Second)
+
+	var (
+		db           *sql.DB
+		err          error
+		retryCounter int
+	)
+
+monitoringLoop:
+	for {
+		select {
+		case <-ticker.C:
+			if err = golf.Conn.Ping(); err != nil {
+				if retryCounter < retry {
+					for i := 0; i < retry; i++ {
+						db, err = establishDBConnection(golf.Logger, c)
```

Comment:
What if clocks ticks between the retries, Should we increase clocking ticking duration a bit?

Reply:
no need for more clarity excute the below code 

`


func  main() {

	t := time.NewTicker(time.Second * 1)

	for {
		select {
		case <-t.C:
			fmt.Println("hii")
		}

	}

	for {
		select {
		case <-t.C:
			fmt.Println("hii")
			time.Sleep(time.Second * 5)
		}
	}
}`

execute the for loops separately to understand better